sbin/hammer/hammer.8
.\" Copyright (c) 2007 The DragonFly Project.  All rights reserved.
.\"
.\" This code is derived from software contributed to The DragonFly Project
.\" by Matthew Dillon <dillon@backplane.com>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\" 3. Neither the name of The DragonFly Project nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific, prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
.\" FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
.\" COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY OR CONSEQUENTIAL DAMAGES (INCLUDING,
.\" BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
.\" OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd December 31, 2017
.Dt HAMMER 8
.Os
.Sh NAME
.Nm hammer
.Nd HAMMER file system utility
.Sh SYNOPSIS
.Nm
.Fl h
.Nm
.Op Fl 2ABFqrvXy
.Op Fl b Ar bandwidth
.Op Fl C Ar cachesize Ns Op Ns Cm \&: Ns Ar readahead
.Op Fl R Ar restrictcmd
.Op Fl T Ar restrictpath
.Op Fl c Ar cyclefile
.Op Fl e Ar scoreboardfile
.Op Fl f Ar blkdevs
.\" .Op Fl s Ar linkpath
.Op Fl i Ar delay
.Op Fl p Ar ssh-port
.Op Fl S Ar splitsize
.Op Fl t Ar seconds
.Op Fl m Ar memlimit
.Ar command
.Op Ar argument ...
.Sh DESCRIPTION
This manual page documents the
.Nm
utility which provides miscellaneous functions related to managing a
.Nm HAMMER
file system.
For a general introduction to the
.Nm HAMMER
file system, its features, and
examples on how to set up and maintain one, see
.Xr HAMMER 5 .
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl 2
Tell the mirror commands to use a 2-way protocol, which allows
automatic negotiation of transaction id ranges.
This option is automatically enabled by the
.Cm mirror-copy
command.
.It Fl A
Make per PFS commands perform on all PFSs if possible.
If the command supports this option, it targets all PFSs of the
.Nm HAMMER
filesystem that the
.Ar filesystem
argument (of that command) belongs to.
Currently
.Cm rebalance ,
.Cm reblock ,
.Cm reblock-btree ,
.Cm reblock-inodes ,
.Cm reblock-dirs
and
.Cm reblock-data
commands support this option.
If the command does not support this option, it does nothing.
.It Fl B
Bulk transfer.
.Cm Mirror-stream
will not attempt to break-up large initial bulk transfers into smaller
pieces.
This can save time but if the link is lost in the middle of the
initial bulk transfer you will have to start over from scratch.
For more information see the
.Fl S
option.
.It Fl b Ar bandwidth
Specify a bandwidth limit in bytes per second for mirroring streams.
This option is typically used to prevent batch mirroring operations from
loading down the machine.
The bandwidth may be suffixed with
.Cm k , m ,
or
.Cm g
to specify values in kilobytes, megabytes, and gigabytes per second.
If no suffix is specified, bytes per second is assumed.
.Pp
Unfortunately this is only applicable to the pre-compression bandwidth
when compression is used, so a better solution would probably be to
use a
.Xr ipfw 8
pipe or a
.Xr pf 4
queue.
.It Fl C Ar cachesize Ns Op Ns Cm \&: Ns Ar readahead
Set the memory cache size for any raw
.Tn I/O .
The default is 16MB.
A suffix of
.Cm k
for kilobytes and
.Cm m
for megabytes is allowed,
else the cache size is specified in bytes.
.Pp
The read-behind/read-ahead defaults to 4
.Nm HAMMER
blocks.
.Pp
This option is typically only used with diagnostic commands
as kernel-supported commands will use the kernel's buffer cache.
.It Fl R Ar restrictcmd
This option is used by hammer ssh-remote to restrict the command later
on in the argument list.
Multiple commands may be specified, separated by a comma (all one argument).
.It Fl T Ar restrictpath
This option is used by hammer ssh-remote to restrict the filesystem path
specified later on in the argument list.
.It Fl c Ar cyclefile
When pruning, rebalancing or reblocking you can tell the utility
to start at the object id stored in the specified file.
If the file does not exist
.Nm
will start at the beginning.
If
.Nm
is told to run for a specific period of time
.Pq Fl t
and is unable to complete the operation it will write out
the current object id so the next run can pick up where it left off.
If
.Nm
runs to completion it will delete
.Ar cyclefile .
.It Fl e Ar scoreboardfile
Update scoreboard file with progress, primarily used by mirror-stream.
.It Fl F
Force operation.
E.g.\&
.Cm cleanup
will not check that time period has elapsed if this option is given.
.It Fl f Ar blkdevs
Specify the volumes making up a
.Nm HAMMER
file system.
.Ar Blkdevs
is a colon-separated list of devices, each specifying a
.Nm HAMMER
volume.
.It Fl h
Show usage.
.It Fl i Ar delay
Specify delay in seconds for
.Cm mirror-read-stream .
When maintaining a streaming mirroring this option specifies the
minimum delay after a batch ends before the next batch is allowed
to start.
The default is five seconds.
.It Fl m Ar memlimit
Specify the maximum amount of memory
.Nm
will allocate during a dedup pass.
Specify a suffix 'm', 'g', or 't' for megabytes, gigabytes, or terabytes.
By default
.Nm
will allocate up to 1G of ram to hold CRC/SHA tables while running dedup.
When the limit is reached the dedup code restricts the range of CRCs to
keep memory use within bounds and runs multiple passes as necessary until
the entire filesystem has been deduped.
.It Fl p Ar ssh-port
Pass the
.Fl p Ar ssh-port
option to
.Xr ssh 1
when using a remote
specification for the source and/or destination.
.It Fl q
Decrease verboseness.
May be specified multiple times.
.It Fl r
Specify recursion for those commands which support it.
.It Fl S Ar splitsize
Specify the bulk splitup size in bytes for mirroring streams.
When a
.Cm mirror-stream
is first started
.Nm
will do an initial run-through of the data to calculate good
transaction ids to cut up the bulk transfers, creating
restart points in case the stream is interrupted.
If we don't do this and the stream is interrupted it might
have to start all over again.
The default is a
.Ar splitsize
of 4GB.
.Pp
At the moment the run-through is disk-bandwidth-heavy but some
future version will limit the run-through to just the B-Tree
records and not the record data.
.Pp
The splitsize may be suffixed with
.Cm k , m ,
or
.Cm g
to specify values in kilobytes, megabytes, or gigabytes.
If no suffix is specified, bytes is assumed.
.Pp
When mirroring very large filesystems the minimum recommended
split size is 4GB.
A small split size may wind up generating a great deal of overhead
but very little actual incremental data and is not recommended.
.It Fl t Ar seconds
Specify timeout in seconds.
When pruning, rebalancing, reblocking or mirror-reading
you can tell the utility to stop after a certain period of time.
A value of 0 means unlimited.
This option is used along with the
.Fl c Ar cyclefile
option to prune, rebalance or reblock incrementally.
.It Fl v
Increase verboseness.
May be specified multiple times.
.It Fl X
Enable compression for any remote ssh specifications.
This option is typically used with the mirroring directives.
.It Fl y
Force
.Dq yes
for interactive questions.
.El
.Pp
The commands are as follows:
.Bl -tag -width indent
.\" ==== synctid ====
.It Cm synctid Ar filesystem Op Cm quick
Generate a guaranteed, formal 64-bit transaction id representing the
current state of the specified
.Nm HAMMER
file system.
The file system will be synced to the media.
.Pp
If the
.Cm quick
keyword is specified the file system will be soft-synced, meaning that a
crash might still undo the state of the file system as of the transaction
id returned but any new modifications will occur after the returned
transaction id as expected.
.Pp
This operation does not create a snapshot.
It is meant to be used
to track temporary fine-grained changes to a subset of files and
will only remain valid for
.Ql @@
access purposes for the
.Cm prune-min
period configured for the PFS.
If you desire a real snapshot then the
.Cm snapq
directive may be what you are looking for.
.\" ==== bstats ====
.It Cm bstats Op Ar interval
Output
.Nm HAMMER
B-Tree statistics until interrupted.
Pause
.Ar interval
seconds between each display.
The default interval is one second.
.\" ==== iostats ====
.It Cm iostats Op Ar interval
Output
.Nm HAMMER
.Tn I/O
statistics until interrupted.
Pause
.Ar interval
seconds between each display.
The default interval is one second.
.\" ==== stats ====
.It Cm stats Op Ar interval
Output
.Nm HAMMER
B-Tree and
.Tn I/O
statistics until interrupted.
Pause
.Ar interval
seconds between each display.
The default interval is one second.
.\" ==== history ====
.It Cm history Ns Oo Cm @ Ns Ar offset Ns Oo Cm \&, Ns Ar length Oc Oc Ar path Ns Oo Cm @ Ns Ar offset Ns Oo Cm \&, Ns Ar length Oc Oc Ar ...
Show the modification history for inode and data of
.Nm HAMMER
files.
If
.Ar offset
is given history is shown for data block at given offset,
otherwise history is shown for inode.
If
.Fl v
is specified
.Ar length
data bytes at given offset are dumped for each version,
default is 32.
.Pp
For each
.Ar path
this directive shows object id and sync status,
and for each object version it shows transaction id and time stamp.
Files has to exist for this directive to be applicable,

sbin/hammer/hammer.c
/*
 * Copyright (c) 2007 The DragonFly Project.  All rights reserved.
 *
 * This code is derived from software contributed to The DragonFly Project
 * by Matthew Dillon <dillon@backplane.com>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 3. Neither the name of The DragonFly Project nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific, prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "hammer.h"

#include <fstab.h>

static __inline void hammer_parse_blkdevs(const char *blkdevs, int oflags);
static void __hammer_parse_blkdevs(const char *blkdevs, int oflags,
		int verify_volume, int verify_count);
static void sigalrm(int signo);
static void sigintr(int signo);
static void usage(int exit_code) __dead2;

int RecurseOpt;
int VerboseOpt;
int QuietOpt;
int NoSyncOpt;
int TwoWayPipeOpt;
int TimeoutOpt;
int DelayOpt = 5;
char *SshPort;
int ForceYesOpt;
int CompressOpt;
int ForceOpt;
int RunningIoctl;
int DidInterrupt;
int BulkOpt;
int AllPFS;
uint64_t BandwidthOpt;
uint64_t SplitupOpt = 4ULL * 1024ULL * 1024ULL * 1024ULL;
uint64_t MemoryLimit = 1024LLU * 1024 * 1024;
const char *SplitupOptStr;
const char *CyclePath;

int
main(int ac, char **av)
{
	char *blkdevs = NULL;
	char *ptr;
	char *restrictcmd = NULL;
	int ch;

	while ((ch = getopt(ac, av,
			    "b:c:de:hf:i:m:p:qrt:v2yABC:FR:S:T:X")) != -1) {
		switch(ch) {
		case '2':
			TwoWayPipeOpt = 1;
			break;
		case 'y':
			ForceYesOpt = 1;
			break;
		case 'b':
			BandwidthOpt = strtoull(optarg, &ptr, 0);
			switch(*ptr) {
			case 'g':
			case 'G':
				BandwidthOpt *= 1024;
				/* fall through */
			case 'm':
			case 'M':
				BandwidthOpt *= 1024;
				/* fall through */
			case 'k':
			case 'K':
				BandwidthOpt *= 1024;
				break;
			case '\0':
				/* bytes per second if no suffix */
				break;
			default:
				usage(1);
				/* not reached */
			}
			break;
		case 'S':
			SplitupOptStr = strdup(optarg);
			SplitupOpt = strtoull(optarg, &ptr, 0);
			switch(*ptr) {
			case 'g':
			case 'G':
				SplitupOpt *= 1024;
				/* fall through */
			case 'm':
			case 'M':
				SplitupOpt *= 1024;
				/* fall through */
			case 'k':
			case 'K':
				SplitupOpt *= 1024;
				break;
			case '\0':
				/* bytes per second if no suffix */
				break;
			default:
				usage(1);
				/* not reached */
			}
			break;
		case 'c':
			CyclePath = optarg;
			break;
		case 'd':
			++DebugOpt;
			break;
		case 'e':
			ScoreBoardFile = optarg;
			break;
		case 'h':
			usage(0);
			/* not reached */
		case 'i':
			DelayOpt = strtol(optarg, NULL, 0);
			break;
		case 'm':
			MemoryLimit = strtouq(optarg, &ptr, 0);
			switch(*ptr) {
			case 't':
			case 'T':
				MemoryLimit *= 1024;
				/* fall through */
			case 'g':
			case 'G':
				MemoryLimit *= 1024;
				/* fall through */
			case 'm':
			case 'M':
				MemoryLimit *= 1024;
				/* fall through */
			case 'k':
			case 'K':
				MemoryLimit *= 1024;
				/* fall through */
			default:
				break;
			}

			/* minimum limit */
			if (MemoryLimit < 1024 * 1024)
				MemoryLimit = 1024 * 1024;
			break;
		case 'p':
			SshPort = optarg;
			break;
		case 'r':
			RecurseOpt = 1;
			break;
		case 'f':
			blkdevs = optarg;
			break;
		case 't':
			TimeoutOpt = strtol(optarg, NULL, 0);
			break;
		case 'v':
			if (QuietOpt > 0)
				--QuietOpt;
			else
				++VerboseOpt;
			break;
		case 'q':
			if (VerboseOpt > 0)
				--VerboseOpt;
			else
				++QuietOpt;
			break;
		case 'A':
			AllPFS = 1;
			break;
		case 'B':
			BulkOpt = 1;
			break;
		case 'C':
			if (hammer_parse_cache_size(optarg) == -1) {
				usage(1);
				/* not reached */
			}
			break;
		case 'F':
			ForceOpt = 1;
			break;
		case 'R':
			if (restrictcmd == NULL)
				restrictcmd = optarg;
			break;
		case 'T':
			if (RestrictTarget == NULL)
				RestrictTarget = optarg;
			break;
		case 'X':
			CompressOpt = 1;
			break;
		default:
			usage(1);
			/* not reached */
		}
	}
	ac -= optind;
	av += optind;
	if (ac < 1) {
		usage(1);
		/* not reached */
	}

	signal(SIGALRM, sigalrm);
	signal(SIGINT, sigintr);

	/*
	 * Check command restriction (used by hammer ssh-remote).  Several
	 * commands may be iterated with a comma.
	 */
	if (restrictcmd) {
		char *elm, *dup;

		dup = ptr = strdup(restrictcmd);
		while ((elm = strsep(&ptr, ",")) != NULL) {
			if (strcmp(av[0], elm) == 0)
				break;
		}
		if (elm == NULL) {
			errx(1, "hammer-remote: request does not match "
				"restricted command");
			/* not reached */
		}
		free(dup);
	}

	/*
	 * Lookup the filesystem type
	 */
	if (hammer_uuid_name_lookup(&Hammer_FSType, HAMMER_FSTYPE_STRING)) {
		errx(1, "uuids file does not have the DragonFly "
			"HAMMER filesystem type");
		/* not reached */
	}

	/*
	 * Parse commands
	 */
	if (strcmp(av[0], "synctid") == 0) {
		hammer_cmd_synctid(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "namekey2") == 0) {
		int64_t key;
		int32_t crcx;
		int len;
		const char *aname = av[1];

		if (aname == NULL) {
			usage(1);
			/* not reached */
		}
		len = strlen(aname);
		key = (uint32_t)crc32(aname, len) & 0xFFFFFFFEU;

		switch(len) {
		default:
			crcx = crc32(aname + 3, len - 5);
			crcx = crcx ^ (crcx >> 6) ^ (crcx >> 12);
			key |= (int64_t)(crcx & 0x3F) << 42;
			/* fall through */
		case 5:
		case 4:
			/* fall through */
		case 3:
			key |= ((int64_t)(aname[2] & 0x1F) << 48);
			/* fall through */
		case 2:
			key |= ((int64_t)(aname[1] & 0x1F) << 53) |
			       ((int64_t)(aname[len-2] & 0x1F) << 37);
			/* fall through */
		case 1:
			key |= ((int64_t)(aname[0] & 0x1F) << 58) |
			       ((int64_t)(aname[len-1] & 0x1F) << 32);
			/* fall through */
		case 0:
			break;
		}
		if (key == 0)
			key |= 0x100000000LL;
		printf("0x%016jx\n", (uintmax_t)key);
		exit(0);
	}
	if (strcmp(av[0], "namekey1") == 0) {
		int64_t key;

		if (av[1] == NULL) {
			usage(1);
			/* not reached */
		}
		key = (int64_t)(crc32(av[1], strlen(av[1])) & 0x7FFFFFFF) << 32;
		if (key == 0)
			key |= 0x100000000LL;
		printf("0x%016jx\n", (uintmax_t)key);
		exit(0);
	}
	if (strcmp(av[0], "namekey32") == 0) {
		int32_t key;

		if (av[1] == NULL) {
			usage(1);
			/* not reached */
		}
		key = crc32(av[1], strlen(av[1])) & 0x7FFFFFFF;
		if (key == 0)
			++key;
		printf("0x%08x\n", key);
		exit(0);
	}
	if (strcmp(av[0], "pfs-status") == 0) {
		hammer_cmd_pseudofs_status(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "pfs-master") == 0) {
		hammer_cmd_pseudofs_create(av + 1, ac - 1, 0);
		exit(0);
	}
	if (strcmp(av[0], "pfs-slave") == 0) {
		hammer_cmd_pseudofs_create(av + 1, ac - 1, 1);
		exit(0);
	}
	if (strcmp(av[0], "pfs-update") == 0) {
		hammer_cmd_pseudofs_update(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "pfs-upgrade") == 0) {
		hammer_cmd_pseudofs_upgrade(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "pfs-downgrade") == 0) {
		hammer_cmd_pseudofs_downgrade(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "pfs-destroy") == 0) {
		hammer_cmd_pseudofs_destroy(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "prune") == 0) {
		hammer_cmd_softprune(av + 1, ac - 1, 0);
		exit(0);
	}
	if (strcmp(av[0], "prune-everything") == 0) {
		hammer_cmd_softprune(av + 1, ac - 1, 1);
		exit(0);
	}
	if (strcmp(av[0], "config") == 0) {
		hammer_cmd_config(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "viconfig") == 0) {
		hammer_cmd_viconfig(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "cleanup") == 0) {
		hammer_cmd_cleanup(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "abort-cleanup") == 0) {
		hammer_cmd_abort_cleanup(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "info") == 0) {
		hammer_cmd_info(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "ssh-remote") == 0) {
		if (ac != 3) {
			usage(1);
			/* not reached */
		}
		hammer_cmd_sshremote(av[1], av[2]);
		exit(0);
	}
	if (strcmp(av[0], "snap") == 0) {
		hammer_cmd_snap(av + 1, ac - 1, 0, 1);
		exit(0);
	}
	if (strcmp(av[0], "snaplo") == 0) {
		hammer_cmd_snap(av + 1, ac - 1, 0, 0);
		exit(0);
	}
	if (strcmp(av[0], "snapq") == 0) {
		hammer_cmd_snap(av + 1, ac - 1, 1, 0);
		exit(0);
	}
	if (strcmp(av[0], "snapls") == 0) {
		hammer_cmd_snapls(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "snaprm") == 0) {
		hammer_cmd_snaprm(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "snapshot") == 0) {
		hammer_cmd_snapshot(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "bstats") == 0) {
		hammer_cmd_bstats(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "iostats") == 0) {
		hammer_cmd_iostats(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "stats") == 0) {
		hammer_cmd_stats(av + 1, ac - 1);
		exit(0);
	}

	if (strncmp(av[0], "history", 7) == 0) {
		hammer_cmd_history(av[0] + 7, av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "rebalance") == 0) {
		signal(SIGINT, sigalrm);
		hammer_cmd_rebalance(av + 1, ac - 1);
		exit(0);
	}
	if (strncmp(av[0], "reblock", 7) == 0) {
		signal(SIGINT, sigalrm);
		if (strcmp(av[0], "reblock") == 0)
			hammer_cmd_reblock(av + 1, ac - 1, HAMMER_IOC_DO_FLAGS);
		else if (strcmp(av[0], "reblock-btree") == 0)
			hammer_cmd_reblock(av + 1, ac - 1, HAMMER_IOC_DO_BTREE);
		else if (strcmp(av[0], "reblock-inodes") == 0)
			hammer_cmd_reblock(av + 1, ac - 1, HAMMER_IOC_DO_INODES);
		else if (strcmp(av[0], "reblock-dirs") == 0)
			hammer_cmd_reblock(av + 1, ac - 1, HAMMER_IOC_DO_DIRS);
		else if (strcmp(av[0], "reblock-data") == 0)
			hammer_cmd_reblock(av + 1, ac - 1, HAMMER_IOC_DO_DATA);
		else {
			usage(1);
			/* not reached */
		}
		exit(0);
	}
	if (strncmp(av[0], "mirror", 6) == 0) {
		if (strcmp(av[0], "mirror-read") == 0)
			hammer_cmd_mirror_read(av + 1, ac - 1, 0);
		else if (strcmp(av[0], "mirror-read-stream") == 0)
			hammer_cmd_mirror_read(av + 1, ac - 1, 1);
		else if (strcmp(av[0], "mirror-write") == 0)
			hammer_cmd_mirror_write(av + 1, ac - 1);
		else if (strcmp(av[0], "mirror-copy") == 0)
			hammer_cmd_mirror_copy(av + 1, ac - 1, 0);
		else if (strcmp(av[0], "mirror-stream") == 0)
			hammer_cmd_mirror_copy(av + 1, ac - 1, 1);
		else if (strcmp(av[0], "mirror-dump") == 0)
			hammer_cmd_mirror_dump(av + 1, ac - 1);
		else {
			usage(1);
			/* not reached */
		}
		exit(0);
	}
	if (strcmp(av[0], "dedup-simulate") == 0) {
		hammer_cmd_dedup_simulate(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "dedup") == 0) {
		hammer_cmd_dedup(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "version") == 0) {
		hammer_cmd_get_version(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "version-upgrade") == 0) {
		hammer_cmd_set_version(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "volume-add") == 0) {
		hammer_cmd_volume_add(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "volume-del") == 0) {
		hammer_cmd_volume_del(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "volume-list") == 0) {
		hammer_cmd_volume_list(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "volume-blkdevs") == 0) {
		hammer_cmd_volume_blkdevs(av + 1, ac - 1);
		exit(0);
	}

	if (strcmp(av[0], "show") == 0) {
		const char *arg = NULL;
		char *p, *dup;
		int filter = -1;
		int obfuscate = 0;
		int indent = 0;

		hammer_parse_blkdevs(blkdevs, O_RDONLY);
		if (ac > 3) {
			errx(1, "Too many options specified");
			/* not reached */
		}
		if (ac > 1)
			arg = av[1];
		if (ac > 2) {
			dup = ptr = strdup(av[2]);
			while ((p = strsep(&ptr, ",")) != NULL) {
				if (strcmp(p, "filter") == 0)
					filter = 1;
				else if (strcmp(p, "nofilter") == 0)
					filter = 0;
				else if (strcmp(p, "obfuscate") == 0)
					obfuscate = 1;
				else if (strcmp(p, "indent") == 0)
					indent = 1;
			}
			free(dup);
		}
		hammer_cmd_show(arg, filter, obfuscate, indent);
		exit(0);
	}
	if (strcmp(av[0], "show-undo") == 0) {
		hammer_parse_blkdevs(blkdevs, O_RDONLY);
		hammer_cmd_show_undo();
		exit(0);
	}
	if (strcmp(av[0], "recover") == 0) {
		__hammer_parse_blkdevs(blkdevs, O_RDONLY, 0, 1);
		hammer_cmd_recover(av + 1, ac - 1);
		exit(0);
	}
	if (strcmp(av[0], "blockmap") == 0) {
		hammer_parse_blkdevs(blkdevs, O_RDONLY);
		hammer_cmd_blockmap();
		exit(0);
	}
	if (strcmp(av[0], "checkmap") == 0) {
		hammer_parse_blkdevs(blkdevs, O_RDONLY);
		hammer_cmd_checkmap();
		exit(0);
	}
	if (strcmp(av[0], "strip") == 0) {
		__hammer_parse_blkdevs(blkdevs, O_RDWR, 0, 0);
		hammer_cmd_strip();
		exit(0);
	}

	usage(1);
	/* not reached */
	return(0);
}

/*
 * Parse the device specification.
 *
 * Multi-volume hammer devices are colon-separated.  Each element
 * may be further expanded via /etc/devtab.  One may also specify
 * a single element which is expanded into multiple elements via
 * /etc/devtab.
 */
static
void
__hammer_parse_blkdevs(const char *blkdevs, int oflags, int verify_volume,
	int verify_count)
{
	volume_info_t volume = NULL;
	char *copy;
	char *volname;
	int vol_count = 0;

	if (blkdevs == NULL) {
		errx(1, "A -f blkdevs specification is required "
			"for this command");
		/* not reached */
	}

	copy = strdup(blkdevs);
	while ((volname = copy) != NULL) {
		if ((copy = strchr(copy, ':')) != NULL)
			*copy++ = 0;
		volname = getdevpath(volname, 0);
		if (strchr(volname, ':')) {
			__hammer_parse_blkdevs(volname, oflags, verify_volume,
				verify_count);
		} else {
			volume = load_volume(volname, oflags, verify_volume);
			assert(volume);
			++vol_count;
		}
		free(volname);
	}
	free(copy);

	assert(volume);
	if (verify_count) {
		if (vol_count != volume->ondisk->vol_count) {
			errx(1, "Volume header says %d volumes, but %d specified.",
				volume->ondisk->vol_count, vol_count);
			/* not reached */
		}
		if (get_root_volume() == NULL) {
			errx(1, "No root volume found");
			/* not reached */
		}
	}
}

static __inline
void
hammer_parse_blkdevs(const char *blkdevs, int oflags)
{
	__hammer_parse_blkdevs(blkdevs, oflags, 1, 1);
}

static
void
sigalrm(int signo __unused)
{
	/* do nothing (interrupts HAMMER ioctl) */
}

static
void
sigintr(int signo __unused)
{
	if (RunningIoctl == 0)
		_exit(1);
	DidInterrupt = 1;
	/* do nothing (interrupts HAMMER ioctl) */
}

static
void
usage(int exit_code)
{
	fprintf(stderr,
		"hammer -h\n"
		"hammer [-2ABFqrvXy] [-b bandwidth] [-C cachesize[:readahead]] \n"
		"       [-R restrictcmd] [-T restrictpath] [-c cyclefile]\n"
		"       [-e scoreboardfile] [-f blkdevs] [-i delay] [-p ssh-port]\n"
		"       [-S splitsize] [-t seconds] [-m memlimit] command [argument ...]\n"
		"hammer synctid <filesystem> [quick]\n"
		"hammer bstats [interval]\n"
		"hammer iostats [interval]\n"
		"hammer stats [interval]\n"
		"hammer history[@offset[,len]] <file> ...\n"
		"hammer namekey1 <path>\n"
		"hammer namekey2 <path>\n"
		"hammer namekey32 <path>\n"
		"hammer cleanup [<filesystem> ...]\n"
		"hammer abort-cleanup\n"
		"hammer info [<dirpath> ...]\n"
		"hammer snapshot [<filesystem>] <snapshot-dir>\n"
		"hammer snapshot <filesystem> <snapshot-dir> [<note>]\n"
		"hammer prune <softlink-dir>\n"
		"hammer prune-everything <filesystem>\n"
		"hammer rebalance <filesystem> [saturation_percentage]\n"
		"hammer reblock[-btree|-inodes|-dirs|-data] "
			"<filesystem> [fill_percentage]\n"
		"hammer pfs-status <dirpath> ...\n"
		"hammer pfs-master <dirpath> [options]\n"
		"hammer pfs-slave <dirpath> [options]\n"
		"hammer pfs-update <dirpath> [options]\n"
		"hammer pfs-upgrade <dirpath>\n"
		"hammer pfs-downgrade <dirpath>\n"
		"hammer pfs-destroy <dirpath>\n"
		"hammer mirror-read <filesystem> [begin-tid]\n"
		"hammer mirror-read-stream <filesystem> [begin-tid]\n"
		"hammer mirror-write <filesystem>\n"
		"hammer mirror-dump [header]\n"
		"hammer mirror-copy [[user@]host:]<filesystem>"
				  " [[user@]host:]<filesystem>\n"
		"hammer mirror-stream [[user@]host:]<filesystem>"
				    " [[user@]host:]<filesystem>\n"
		"hammer ssh-remote command filesystem\n"
		"hammer version <filesystem>\n"
		"hammer version-upgrade <filesystem> <version> [force]\n"
		"hammer volume-add <device> <filesystem>\n"
		"hammer volume-del <device> <filesystem>\n"
		"hammer volume-list <filesystem>\n"
		"hammer volume-blkdevs <filesystem>\n"
	);

	fprintf(stderr, "\nHAMMER utility version 3+ commands:\n");

	fprintf(stderr,
		"hammer config [<filesystem> [<configfile>]]\n"
		"hammer viconfig [<filesystem>]\n"
		"hammer snap <path> [<note>]\n"
		"hammer snaplo <path> [<note>]\n"
		"hammer snapq <dir> [<note>]\n"
		"hammer snaprm <path> ...\n"
		"hammer snaprm <transid> ...\n"
		"hammer snaprm <filesystem> <transid> ...\n"
		"hammer snapls [<path> ...]\n"
	);

	fprintf(stderr, "\nHAMMER utility version 4+ commands:\n");

	fprintf(stderr,
		"hammer -f blkdevs blockmap\n"
		"hammer -f blkdevs checkmap\n"
		"hammer -f blkdevs [-qqq] show [lo:objid]\n"
		"hammer -f blkdevs show-undo\n"
		"hammer -f blkdevs recover <target_dir> [full|quick]\n"
		"hammer -f blkdevs strip\n"
	);

	fprintf(stderr, "\nHAMMER utility version 5+ commands:\n");

	fprintf(stderr,
		"hammer dedup-simulate <filesystem>\n"
		"hammer dedup <filesystem>\n"
	);

	exit(exit_code);
}

